src/tactic/apply_fun.lean

```diff
@@ -0,0 +1,93 @@
+import tactic.monotonicity
+import order.basic
+
+open tactic interactive (parse) interactive (loc.ns)
+     interactive.types (texpr location) lean.parser (tk)
+
+local postfix `?`:9001 := optional
+
+meta def apply_fun_name (e : expr) (h : name) (M : option pexpr) : tactic unit :=
+do {
+  H ← get_local h,
+  t ← infer_type H,
+  match t with
+  | `(%%l = %%r) := do
+      tactic.interactive.«have» h none ``(congr_arg %%e %%H),
+      tactic.clear H
+  | `(%%l ≤ %%r) := do
+       if M.is_some then do
+         Hmono ← M >>= tactic.i_to_expr,
+         tactic.interactive.«have» h none ``(%%Hmono %%H)
+       else do {
+         n ← get_unused_name `mono,
+         tactic.interactive.«have» n ``(monotone %%e) none,
+         do { intro_lst [`x, `y, `h], `[dsimp, mono], skip } <|> swap,
+         Hmono ← get_local n,
+         tactic.interactive.«have» h none ``(%%Hmono %%H) },
+       tactic.clear H
+  | _ := skip
+  end,
+  -- let's try to force β-reduction at `h`
+  try (tactic.interactive.dsimp tt [] [] (loc.ns [h])
+         {eta := false, beta := true})
+  } <|> fail ("failed to apply " ++ to_string e ++ " at " ++ to_string h)
+
+namespace tactic.interactive
+/-- Apply a function to some local assumptions which are either equalities or
+    inequalities. For instance, if the context contains `h : a = b` and
+    some function `f` then `apply_fun f at h` turns `h` into `h : f a = f b`.
+    When the assumption is an inequality `h : a ≤ b`, a side goal `monotone f`
+    is created, unless this condition is provided using
+    `apply_fun f at h using P` where `P : monotone f`, or the `mono` tactic can
+    prove it. -/
+meta def apply_fun (q : parse texpr) (locs : parse location)
+  (lem : parse (tk "using" *> texpr)?) : tactic unit :=
+do e ← tactic.i_to_expr q,
```

Comment:
I think you should pass `q` to `apply_fun_name` as a `pexpr` and save the elaboration for later. Right now, you're elaborating `q` out of context. You won't be able to use `apply_fun id`, for example, or `apply_fun (λ n, n+1)` on an integer equality.

Reply:
I hate pretending I understand tactic writing, but the test on line 74 proves your example is wrong. But again I'd be happy to change this. We can probably do that together tomorrow.

Reply:
That test is a nat equality, not an int equality. Lean will decide that `(λ n, n+1) ` is `nat -> nat` by default, since in the absence of any other information, `1` is a nat. If you change `a` and `b` in that example to `int`s, it should fail, since Lean will still decide the function is `nat -> nat`.